Feature: Implement CREATE & UPDATE food functionalities

nick-y-ito opened the pull request:
<!--

Title Naming Conventions:

- Start with a capitalized prefix followed by an imperative form of a verb.
- Refer to the "Types of Branches" section for the prefix:
  https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#types-of-branches

Example:
- Feature: Add button to the home page
- Fix: Resolve error in data fetching
- Hotfix: Fix user authentication issue
- Refactor: Reorganize the file structure
- Test: Add unit tests for form validation
- Docs: Update installation guide

-->

## Overview

<!-- Provide a brief description of the changes introduced by this PR. -->

This PR implements the CRATE & UPDATE food functionalities.

## Changes

<!-- List the changes -->
<!-- Delete this section if not needed -->

- Installed the npm package [`result-ts-type`](https://www.npmjs.com/package/result-ts-type) that @hitohata created, and utilized it for actions and API client.
- Created submit handlers in the `actions.ts` and added unit test
- Added `alert()`, which should be replaced with the [`toast`](#205)
- Validations might not be perfectly refined
- Refactored the `request` function for each client/server components to be able to handle empty a response body, for example, in case the status code is 204
- Refactored the unit test code for the `request` function for each client/server components

## Review points

<!-- List the points you would like to be reviewed in detail and the points you are not confident about. -->
<!-- Delete this section if not needed -->

- Please point out everything you think could be easier to read.
- I did not change any UI, so please focus your review on:
  - The functionality (whether or not the API calls are working) 
  - Readability of the code

## Screen Captures

<!-- Add screenshots or screen recordings of the changes -->
<!-- Necessary for UI changes -->
<!-- Delete this section if not needed -->


https://github.com/nishiki-tech/nishiki-frontend/assets/99148565/b5edc990-bba5-4095-9baf-a5a5ae8f3f4e


## References

<!-- List all the URLs of the references (documents, articles, YouTube videos, etc.) that you used to complete this task -->
<!-- If you have referred to responses from ChatGPT or other AIs, please include the links to the conversations if possible -->
<!-- Delete this section if not needed -->

- NPM package [`result-ts-type`](https://www.npmjs.com/package/result-ts-type) that @hitohata created
- 

## Notes

<!-- Write any note or comment. You can share your thoughts or ideas. -->
<!-- Delete this section if not needed -->

-

## Assignee Checklist:

<!-- You can tick the checkboxes if not applicable -->

- [x] The naming convention of the PR title is correct (See the comment at the top of this template)
- [x] The base branch is correct (See: [Types of Branches](https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#types-of-branches))
- [x] The branch name follows the [Branch Naming Conventions](https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#branch-naming-conventions)
- [x] The correct assignees and reviewers have been designated for this PR
- [x] The coding style follows the [Coding Style Guide](https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [x] All the related issues are associated with this PR
- [x] All criteria in the associated issues are met (please tick the checkboxes)
- [x] My changes do not generate new warnings or errors (especially in the console)

## Reviewer Checklist:

<!-- You can tick the checkboxes if not applicable -->

- [x] Double-check the "Assignee Checklist"
- [x] The code follows the generic best practices and our [Coding Style Guide](https://github.com/nishiki-tech/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [x] The code is well-commented and easy to understand
- [x] The UI changes accurately reflect the provided design

kanta1207 commented on src/lib/api/container/client/containerApiClient.client.ts:
You didn't add test for containetApiClient.client.ts because there's not much things to test?
Indeed, it'll be just to check
- if the request function is successfully called
- if the function returns an error when request function cause an error

Just in case if you're unintentionally missing. Or if you're omitting it intentionally, please kindly share your thoughts 🙏

nick-y-ito replied:
Thank you for reminding me! I had completely slipped my mind.

kanta1207 reviewed (commented):
Functionality and readability overall seemed to be perfect to me!
Rather sorry for not being able to find a room for improvement 😅, it's really looking nice to me.

Just one thing to mention, is it intentional to omit the test for `container.client.ts`?
It's indeed not containing much things to test tho, just in case you missed it unintentionally.

nick-y-ito commented:
@kanta1207,

Thank you for taking the time to review. I have made the updates below based on your feedback:

- Added a unit test for the container's API client.
- Added a unit test for the food's actions.
- Updated the `request` function for the server. *Please note that I had only made modifications to the one for the client in the previous state.
- Updated the unit test for `request` functions for both client and server.

Please review these changes when you have the time. Thank you!

kanta1207 commented on src/features/foods/lib/actions.test.ts:
This one looks good, too!

kanta1207 reviewed (approved):
Thank you for addressing the review points so quickly! I'll approve this PR, thank you for your work :)

nick-y-ito commented:
@kanta1207 
Thank you for your review approval 😄
